server.go

```diff
@@ -376,6 +426,107 @@ func (s *Server) ReadTraceSocket(packetPool *sync.Pool, reuseport bool) {
 	}
 }
 
+func (s *Server) handleTCPGoroutine(conn net.Conn) {
+	defer func() {
+		s.ConsumePanic(recover())
+	}()
+
+	defer func() {
+		err := conn.Close()
+		s.statsd.Count("tcp.disconnects", 1, nil, 1.0)
+		if err != nil {
+			log.WithError(err).Error("TCP close failed")
+		}
+	}()
+	s.statsd.Count("tcp.connects", 1, nil, 1.0)
+
+	if tlsConn, ok := conn.(*tls.Conn); ok {
+		// complete the handshake to verify the certificate
+		if !tlsConn.ConnectionState().HandshakeComplete {
+			err := tlsConn.Handshake()
+			if err != nil {
+				s.statsd.Count("tcp.tls_handshake_failures", 1, nil, 1.0)
+				log.WithError(err).Error("TLS Handshake failed")
+				return
+			}
+		}
+
+		state := tlsConn.ConnectionState()
+		var clientCert pkix.RDNSequence
+		if len(state.PeerCertificates) > 0 {
+			clientCert = state.PeerCertificates[0].Subject.ToRDNSequence()
+		}
+		log.WithFields(logrus.Fields{
+			"peer":        conn.RemoteAddr().String(),
+			"client_cert": clientCert,
+		}).Debug("Starting TLS connection")
+	} else {
+		log.WithFields(logrus.Fields{
+			"peer": conn.RemoteAddr().String(),
+		}).Debug("Starting TCP connection")
+	}
+
+	// Scanner is nearly the same performance as a custom implementation
+	buf := bufio.NewScanner(conn)
+	for buf.Scan() {
+		metric, err := samplers.ParseMetric(buf.Bytes())
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				logrus.ErrorKey: err,
+				"packet":        strconv.QuoteToASCII(buf.Text()),
+			}).Error("Could not parse packet; Closing TCP connection")
+			s.statsd.Count("packet.error_total", 1, []string{"packet_type:metric"}, 1.0)
+			return
+		}
+		s.Workers[metric.Digest%uint32(len(s.Workers))].PacketChan <- *metric
+	}
+	if buf.Err() != nil {
+		log.WithError(buf.Err()).Error("Error reading from TCP client")
+	}
+	log.WithField("peer", conn.RemoteAddr().String()).Debug("Closing TCP connection")
+}
+
+// StartTCP listens on Server.TCPAddr for new connections, starting a goroutine for each.
```

Comment:
Can we update the name of this to be similar to the other `Read...` methods we have for metrics and trace? The relationship isn't really 1:1, but they seem to be pretty analogous.